lib/features/build_server/compiler.dart

```diff
@@ -0,0 +1,143 @@
+import 'dart:io';
+
+import 'package:flyde/core/fs/configs/compiler_config.dart';
+import 'package:flyde/core/fs/compiler/installed_compiler.dart';
+import 'package:flyde/core/fs/wrapper/source_file.dart';
+import 'package:flyde/features/build_server/cache/implementation_object_ref.dart';
+import 'package:flyde/features/build_server/cache/project_cache.dart';
+import 'package:path/path.dart';
+
+class Compiler {
+  CompilerConfig _config;
+
+  final ProjectCache _cache;
+
+  Map<String, String> _projectFiles;
+
+  List<String>? _outdatedFiles;
+
+  Compiler(this._config, this._projectFiles, this._cache);
+
+  /// List of all outdated project file ids.
+  /// These files have to be `insert`ed.
+  Future<List<String>> get outdatedFiles async {
+    if (_outdatedFiles != null) {
+      return _outdatedFiles!;
+    }
+
+    return await _syncCache();
+  }
+
+  Future<File?> get lastExecutable async {
+    try {
+      return await _cache.executable;
+    } catch (_) {
+      return null;
+    }
+  }
+
+  void update(CompilerConfig config, Map<String, String> projectFiles) {
+    _projectFiles = projectFiles;
+    _config = config;
+    _outdatedFiles = null;
+  }
+
+  /// Adds the [file] to the cache. Required for all files which should be compiled.
+  Future<void> insert(SourceFile file) async {
+    await _cache.insert(file);
+  }
+
+  /// Compiles the project.
+  Future<void> compile() async {
+    if (_outdatedFiles == null) {
+      await _syncCache();
+    }
+
+    if (_outdatedFiles!.isNotEmpty) {
+      throw Exception(
+          'There are ${_outdatedFiles!.length} outdated files. Use `insert` to sync your project.');
```

Comment:
Caused by a bug downstream, this will never throw. But when that is resolved, the _outdatedFiles should be deleted in `sync`, that is a little unclear (the doc-comment would let me guess that I have to delete them manually [as I am supposed to be returned a list of all outdated files, which is actually a list of required files])

Reply:
`insert` is part of the sync process as files which are stored as an older version or are completely unavailable are inserted to the cache. I've updated the message to make this a little bit clearer. I'm not sure if you might have misunderstood some parts of the compiler and cache. Maybe the updated `outdadtedFiles` description helps. 

Please feel free to ask questions.

Could you please elaborate on the 'bug downstream' and why this can never throw? There is actually a unit test which ensures this error is thrown and... well... it does not fail ;)

Reply:
This was a misunderstaning, as I read the old `sync`documentation as if the method would return all _deleted_ files, which would always be an empty list due to a bug in the old `sync`.